multistream-2.0/spec.md

```diff
@@ -0,0 +1,412 @@
+This is the subset of identify needed for protocol negotiation.
+
+### Multistream Use
+
+The `multistream/multicodec` protocol is simply two varint multicodecs: the
+multistream-use multicodec followed by the multicodec for the protocol to be
+used. This protocol supports unidirectional streams. If the stream is
+bidirectional, the receiver must acknowledge a successful protocol negotiation
+by responding with the same multistream-use protocol sequence.
+
+Every stream starts with multistream-use. Every other protocol defined here will
+be assigned a multicodec and selected with `multistream/multicodec.`
+
+This protocol should *also* be trivial to optimize in hardware simply by prefix
+matching (i.e., matching on the first N (usually 16-32) bits of the
+stream/message).
+
+### Multistream String
+
+The `multistream/string` protocol is like the `multistream/multicodec` protocol
+*except* that it uses a string to identify the protocol. To do so, the initiator
+simply sends a varint length followed by the name of the protocol.
+
+Including the `multistream/multicodec` portion, the initiator would send:
+
+```
+<multistream/multicodec><multistream/string><length(varint)><name(string)>
+```
+
+Note: This used to use a fixed-width 16 bit number for a length. However, a
+varint *really* isn't going to cost us much, if anything, in terms of
+performance as most protocol names will be <= 128 bytes long. On the other hand,
+using different number formats everywhere *will* cost us in terms of complexity.
+
+### Multistream Dynamic
+
+The `multistream/dynamic` protocol is used to select a protocol using a
+*receiver specified*, session-ephemeral protocol ID. These IDs are analogues of
+ephemeral ports.
+
+In this protocol, the stream initiator sends a varint ID specified by the
+*receiver* to the receiver.
+
+Format:
+
+```
+<multistream/multicodec><multistream/dynamic><id(varint)>
+```
+
+The ID 0 is reserved for saying "same protocol" on a bidirectional stream. The
+receiver of a bidirectional stream can't reuse the same dynamic ID that the
+initiator used as this dynamic ID is relative *to* the receiver. Really, this
+last rule *primarily* exists to side-step the TCP simultaneous connect issue.
+
+This protocol has *also* been designed to be hardware friendly:
+
+1. Hardware can compare the first 16 bits of the message against
+   `<multistream/multicodec><multistream/dynamic>`.
+2. It can then route the message based on the dynamic ID. The fact that these
+   IDs are chosen by the *receiver* means that the receiver can reuse the same
+   IDs for all connected peers (reusing the same hardware routing table).
+   
+### Serial Stream
+
+The `serial-stream` protocol is the simplest possible stream multiplexer.
+Unlike other stream multiplexers, `serial-stream` can only multiplex streams
+in *serial*. That is, it has to close the current stream to open a new one.
+
+The protocol is:
+
+```
+<header (signed 16 bit int)>
```

Comment:
Endianess of this integer should be defined or endianness of all integers in the document should be defined.

At this time I'm assuming network byte order/big-endian.

Reply:
Yes. Network order.